core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/batchutilities/BatchBufferTest.java

```diff
@@ -0,0 +1,297 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.testng.Assert;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.ThreadFactoryBuilder;
+
+public class BatchBufferTest {
+
+    private static final Logger log = Logger.loggerFor(BatchBufferTest.class);
+    private BatchManager<String, String, List<RequestWithId>> buffer;
+    private ScheduledExecutorService scheduledExecutor;
+    private String defaultDestination;
+
+    @Before
+    public void setUp() {
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().threadNamePrefix("batch-buffer").build();
+        scheduledExecutor = Executors.newSingleThreadScheduledExecutor(threadFactory);
+        buffer = new BatchManager<>(10, Duration.ofMillis(200), scheduledExecutor,
+                                    batchingFunction, unpackResponseFunction, getBatchGroupIdFunction);
+        defaultDestination = "dest0";
+    }
+
+    @After
+    public void tearDown() {
+        buffer.close();
+        scheduledExecutor.shutdownNow();
+    }
+
+    @Test
+    public void sendTenRequests() {
+        Map<String, String> requests = createRequestsOfSize(10);
+        Map<String, CompletableFuture<String>> responses = createAndSendResponses(requests);
+        checkAllResponses(requests, responses);
+    }
+
+    @Test
+    public void sendTwentyRequests() {
+        Map<String, String> requests = createRequestsOfSize(20);
+        Map<String, CompletableFuture<String>> responses = createAndSendResponses(requests);
+        checkAllResponses(requests, responses);
+    }
+
+    @Test
+    public void scheduleSendFiveRequests() {
+        Map<String, String> requests = createRequestsOfSize(5);
+
+        long startTime = System.nanoTime();
+        Map<String, CompletableFuture<String>> responses = createAndSendResponses(requests);
+        CompletableFuture.allOf(responses.values().toArray(new CompletableFuture[0])).join();
+        long endTime = System.nanoTime();
+
+        Assert.assertTrue(Duration.ofNanos(endTime - startTime).toMillis() > 200);
+        checkAllResponses(requests, responses);
+    }
+
+    @Test
+    public void cancelScheduledBatch() {
+        Map<String, String> requests = createRequestsOfSize(10);
+
+        long startTime = System.nanoTime();
+        Map<String, CompletableFuture<String>> responses = createAndSendResponses(0, 5, requests);
+        waitForTime(190);
+        responses.putAll(createAndSendResponses(5, 5, requests));
+        CompletableFuture.allOf(responses.values().toArray(new CompletableFuture[0])).join();
+        long endTime = System.nanoTime();
+
+        Assert.assertTrue(Duration.ofNanos(endTime - startTime).toMillis() < 400);
+        checkAllResponses(requests, responses);
+    }
+
+    @Test
+    public void scheduleTwoBatchesToSameDestination() {
```

Comment:
sendTwentyRequests should manually flush the buffer twice, whereas scheduling two batches to the same destination should flush the buffer twice through a timeout (instead of a manual flush triggered from exceeding the maxBatchItems).